New Feature: Implements a modal to edit the Redirect plugin settings when needed

roland-d opened the pull request:
Pull Request for Issue #7259 

### Summary of Changes
This is a redo of the original done at #7259 with some small code changes

### Testing Instructions
1. Make sure that the Redirect plugin is either disabled or the Collection of URLs is disabled
2. Go to administrator/index.php?option=com_redirect
3. When the collection of URLs is disabled you will see this message:
![image](https://user-images.githubusercontent.com/359377/27507070-1d84f518-58c8-11e7-9173-3e285b2fca27.png)

When the plugin is disabled you will see this message:
![image](https://user-images.githubusercontent.com/359377/27507073-354ff846-58c8-11e7-99fa-81df2930d10b.png)

4. Click on the Redirect System Plugin link and you are taken to a new page where you can edit the plugin settings. Now you need to navigate back to the Redirect plugin.
5. Apply patch
6. When the collection of URLs is disabled you will see this message:
![image](https://user-images.githubusercontent.com/359377/27507101-90497966-58c8-11e7-81de-8120f5fceab2.png)

When the plugin is disabled you will see this message:
![image](https://user-images.githubusercontent.com/359377/27507097-7aa874c2-58c8-11e7-8bb6-1659874677c5.png)

I changed the regular link to a button so it is easier to see where to click.
7. Click on the Redirect System Plugin link and you are shown a modal in which you can change the settings. Update the settings and click Save.
8. The page is refreshed and the message has changed.

### Expected result
I stay on the same page and can edit the settings.

### Actual result
I am taking to a new page and need to navigate back.

### Documentation Changes Required
None

Pinging @brianteeman @franz-wohlkoenig for testing :)

roland-d committed:
Redo of #7259 and implements a modal to edit the Redirect plugin settings when needed

franz-wohlkoenig commented:
can you please change Color on hover for easier reading?
![bildschirmfoto 2017-06-24 um 10 43 13](https://user-images.githubusercontent.com/8235763/27507169-06fbaa7e-58ca-11e7-97c9-d9f9e892c75d.jpg)

roland-d commented:
@franz-wohlkoenig I can if you tell me which class I should be using :)

franz-wohlkoenig commented:
i can't, i'm the one who ask;-) CSS-Gurus here?

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 4a8b5f8210496f2207f7cc8626d9e3a19ff26ddb<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

brianteeman commented on administrator/components/com_plugins/views/plugin/tmpl/modal.php:
```diff
@@ -0,0 +1,26 @@
+ * @package     Joomla.Administrator
+ * @subpackage  com_plugins
+ *
+ * @copyright   Copyright (C) 2005 - 2015 Open Source Matters, Inc. All rights reserved.
```
2017

Quy commented:
@roland-d Use the `alert-link` class instead of a button.

brianteeman commented:
@roland-d other than the previouslt mentioned css issue its all good - thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

roland-d committed:
Fixed CSS class and copyright
Changed reload of parent page

roland-d commented:
Thanks everybody for your input. I have made the following changes:
- Fixed the copyright
- Changed the CSS class from btn to alert-link
- Changed the reload of the parent page because the old way showed JS errors and had issues if you filtered the list before clicking the plugin modal

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on cb243fc22c9b2b4a6935fb1f1f07b044207866e6<br /><br />Link is difficult to read.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

roland-d commented:
@franz-wohlkoenig I could remove the label class and then it is just bold text. That just doesn't look clear enough to me that it is a link you can click. @Quy Any other options for styling this so it looks more like a clickable link?

mbabker commented:
Improve the implementation of the `alert-link` class if you don't think it stands out as a link (which is just a backport from Bootstrap 3).  Please don't try to fudge the display with extra UI elements to make it "stand out".

roland-d committed:
Removed label class

roland-d commented:
Ok, I removed the label class. Declaring styling out of scope :)

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 6f1d7aab8e4b2d9918bdee76ac12935c3f6eee03<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

infograf768 commented:
Can't we make this generic? I mean not specific to the redirect plugin?
I am talking about the edit.php
`window.top.setTimeout('window.parent.location = \"index.php?option=com_redirect&view=links\"', 1000);`

We could use the modal for other instances of editing a plugin when we have such situation:
in com_associations for example for the system - languagefilter and in com_fields for system - Fields.

roland-d commented:
@infograf768 This was generic and then I made the mistake of putting that URL in there :P I will change this so it is generic.

dgt41 commented:
@infograf768 in Joomla4 we have [events](https://github.com/joomla/joomla-cms/blob/4.0-dev/media/system/js/core.js#L932-L1021) so this can be a lot nicer and of course not generic.

Also this probably needs some attention when merging static to j4 (we don't have inline scripts anymore)

roland-d committed:
Make location generic

andrepereiradasilva commented on administrator/components/com_redirect/views/links/view.html.php:
```diff
@@ -58,20 +66,30 @@ public function display($tpl = null)
 		{
-			$link = JRoute::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . RedirectHelper::getRedirectPluginId());
-			$app->enqueueMessage(JText::sprintf('COM_REDIRECT_PLUGIN_DISABLED', $link), 'error');
+			$this->redirectPluginId = RedirectHelper::getRedirectPluginId();
```
since you already added the `id` to` JPluginHelper` db query return data, can't you just use `JPluginHelper::getPlugin('system', 'redirect')->id` and remove/deprecate `RedirectHelper::getRedirectPluginId`?

roland-d replied:
You cannot do that because when the plugin is disabled the function will return an empty array.

andrepereiradasilva replied:
Right! Forget that

andrepereiradasilva commented on administrator/components/com_redirect/views/links/tmpl/default.php:
```diff
@@ -23,6 +23,27 @@
+		<?php
+			if ($this->redirectPluginId)
+			{
+				$link = JRoute::_('index.php?option=com_plugins&amp;client_id=0&amp;task=plugin.edit&amp;extension_id=' . $this->redirectPluginId . '&amp;tmpl=component&amp;layout=modal');
```
why use `&amp;` in JRoute::_()?

roland-d replied:
No reason, just copy and paste :)

roland-d committed:
Changed amp usage

franz-wohlkoenig commented:
@roland-d please give a go if PR is ready for Test.

roland-d commented:
@franz-wohlkoenig Let's ask @andrepereiradasilva if he wants anymore changes. If not, we can test.

roland-d commented:
@franz-wohlkoenig Looks like we can test, @andrepereiradasilva has not raised any other issues.

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on 6f5fc467b57bd848d68f6266ea5d66951c973780<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

Quy commented on administrator/components/com_redirect/views/links/view.html.php:
```diff
@@ -58,20 +66,30 @@ public function display($tpl = null)
+
+			if ($this->enabled && !$this->collect_urls_enabled)
+			{
+				$app->enqueueMessage(JText::_('COM_REDIRECT_PLUGIN_ENABLED') . JText::sprintf('COM_REDIRECT_COLLECT_URLS_DISABLED', $link), 'notice');
```
Add a space between sentences.

Quy commented on administrator/components/com_redirect/views/links/view.html.php:
```diff
@@ -58,20 +66,30 @@ public function display($tpl = null)
+				'link',
+				'#plugin' . $this->redirectPluginId . 'Modal',
+				JText::_('COM_REDIRECT_SYSTEM_PLUGIN'),
+				'role="button" class="alert-link" data-toggle="modal" title="' . JText::_('COM_REDIRECT_SYSTEM_PLUGIN') . '" id="title-' . $this->redirectPluginId . '"'
```
Remove `role="button"`.
Remove `title` attribute since it is the same value as the link text.

roland-d commented:
Thanks for the feedback but I am going to wait a week to make changes in case anybody else comes up with something.

zero-24 commented on administrator/components/com_redirect/views/links/tmpl/default.php:
```diff
@@ -23,6 +23,27 @@
 	<div id="j-main-container">
 		<?php echo JLayoutHelper::render('joomla.searchtools.default', array('view' => $this)); ?>
+		<?php
+			if ($this->redirectPluginId)
```
please use opening and closing `<?php ?>` for everyline of PHP in layout files ;)

roland-d replied:
That is just insane. It's a complete block of code, why would you want to open and close tags for every single line?

zero-24 replied:
Just copy:

```php

			<?php if ($this->redirectPluginId) : ?>
				<?php $link = JRoute::_('index.php?option=com_plugins&client_id=0&task=plugin.edit&extension_id=' . $this->redirectPluginId . '&tmpl=component&layout=modal'); ?>
				<?php echo JHtml::_(
					'bootstrap.renderModal',
					'plugin' . $this->redirectPluginId . 'Modal',
					array(
						'url'        => $link,
						'title'      => JText::_('COM_REDIRECT_EDIT_PLUGIN_SETTINGS'),
						'height'     => '400px',
						'modalWidth' => '60',
						'footer'     => '<button class="btn" data-dismiss="modal" aria-hidden="true">'
							. JText::_("JLIB_HTML_BEHAVIOR_CLOSE") . '</button>'
							. '<button class="btn btn-success" data-dismiss="modal" aria-hidden="true" onclick="jQuery(\'#plugin' . $this->redirectPluginId . 'Modal iframe\').contents().find(\'#saveBtn\').click();">'
							. JText::_("JSAVE") . '</button>'
					)
				); ?>
			<?php endif; ?>
```

Quy commented:
In `\administrator\templates\hathor\html\com_redirect\links\default.php` line 148, it is missing an argument.
```
				<span class="enabled"><?php echo JText::_('COM_REDIRECT_COLLECT_URLS_DISABLED'); ?></span>
```

roland-d committed:
Updated after feedback

roland-d commented:
@Quy requested a change in the Hathor override but what we have now is 2 messages. One is the enqueued message and the other the regular displayed message. The enqueued message contains the modal link which doesn't work in Hathor anyway. Are we leaving it like this or add a harcoded check for Hathor? A template which is going to be removed anyway.

zero-24 commented:
Thanks closing

zero-24 commented:
Sorry broken autocorrect

roland-d commented:
Can we move this forward or does it need more changes?

infograf768 commented:
This will not be B/C language wise if you just modify the strings. We get a 404 with non updated language packs (Tested with fa-IR and fr-FR)
I suggest to create new strings/constants and marked the other ones as deprecated by a comment above them.

Obsolete:
; The following string is deprecated and will be removed with 4.0.
COM_REDIRECT_COLLECT_URLS_DISABLED
; The following string is deprecated and will be removed with 4.0.
COM_REDIRECT_PLUGIN_DISABLED

Possible new constants:
COM_REDIRECT_COLLECT_MODAL_URLS_DISABLED
COM_REDIRECT_PLUGIN_MODAL_DISABLED

infograf768 commented:
Also I remarked an RTL issue that was also present before this patch
We should not use
`$app->enqueueMessage(JText::_('COM_REDIRECT_PLUGIN_ENABLED') . ' ' . JText::sprintf('COM_REDIRECT_COLLECT_URLS_DISABLED', $link), 'notice');`

As this is not RTL aware. The first string displays on the left in RTL.

I think we should rather have a sprintf with 2 variables, 1 for the link, the other for the string.
something like:
`$app->enqueueMessage(JText::sprintf('COM_REDIRECT_COLLECT_URLS_DISABLED', JText::_('COM_REDIRECT_PLUGIN_ENABLED'), $link), 'notice');`

`COM_REDIRECT_COLLECT_URLS_DISABLED="%1$s The 'Collect URLs' option in the %2$s is disabled. Error page URLs will not be collected by this component."`

roland-d commented:
@infograf768 
> We get a 404 with non updated language packs (Tested with fa-IR and fr-FR)

You can't get a 404 when using JText. If a language string doesn't exist it will just show the supplied string. What 404 are you talking about?

I will look into the sprintf for RTL

infograf768 commented:
> You can't get a 404 when using JText. If a language string doesn't exist it will just show the supplied string. What 404 are you talking about?

As you have not changed the constant, the original value for the string is
`COM_REDIRECT_PLUGIN_DISABLED="The <a href="_QQ_"%s"_QQ_">Redirect System Plugin</a> is disabled. It needs to be enabled for this component to work."`

Therefore, if the language pack is NOT updated,  % is the `$link` value
and we get a 404 because `<a href="_QQ_"JRoute::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . RedirectHelper::getRedirectPluginId());"_QQ_">Redirect System Plugin</a>` makes no sense.

I suggest you just install French available pack for example + your PR and test.

infograf768 commented:
@roland-d 
If you like, I will make an animated gif

roland-d commented:
@infograf768 No need I think, I get the issue you are pointing out.

roland-d committed:
Updated the language constants and RTL issue

roland-d commented:
@infograf768 I believe it should be correct now, I tested this with both French and English.

Quy commented:
In Hathor, invalid markup `<a href="<a href="index.php?option=com_plugins&filter_search=redirect" >Redirect System Plugin</a>">Redirect System Plugin</a>`

```
<span class="enabled">The 'Collect URLs' option in the <a href="<a href="index.php?option=com_plugins&filter_search=redirect" >Redirect System Plugin</a>">Redirect System Plugin</a> is disabled. Error page URLs will not be collected by this component.</span>
```

roland-d committed:
Fixed invalid markup in Hathor

infograf768 commented:
@roland-d 
Hathor still broken here:
![screen shot 2017-07-03 at 11 36 10](https://user-images.githubusercontent.com/869724/27787072-e269c512-5fe3-11e7-9965-920a5ebbd562.png)

The link in the Error Message looks correct but it does not display the Modal.
I get a js error
`Empty string passed to getElementById().`

The link in the table footer is wrong. It should be a sprintf
line 150 to 152 should be
```
		<?php else : ?>
			<span class="disabled"><?php echo JText::sprintf('COM_REDIRECT_PLUGIN_DISABLED', 'index.php?option=com_plugins&filter_search=redirect'); ?></span>
		<?php endif; ?>
```

roland-d commented:
I don't think there should be a modal in Hathor, should there be?

infograf768 commented:
We use Modals in Hathor, for example when we have associations.
But, yes, I don't think it is necessary, taking into account the fact that we are not supposed to code for hathor anymore. We just should not destroy what works ;) 
The Error message should have the same link as the table footer, i.e. link to the old behavior
`'index.php?option=com_plugins&filter_search=redirect'`

infograf768 commented:
I could get it to work in a basic way though, by adding
```
<?php if ($this->redirectPluginId) : ?>
			<?php $link = JRoute::_('index.php?option=com_plugins&client_id=0&task=plugin.edit&extension_id=' . $this->redirectPluginId . '&tmpl=component&layout=modal'); ?>
			<?php echo JHtml::_(
				'bootstrap.renderModal',
				'plugin' . $this->redirectPluginId . 'Modal',
				array(
					'url'        => $link,
					'title'      => JText::_('COM_REDIRECT_EDIT_PLUGIN_SETTINGS'),
					'height'     => '400px',
					'modalWidth' => '60',
					'footer'     => '<button class="btn" data-dismiss="modal" aria-hidden="true">'
						. JText::_("JLIB_HTML_BEHAVIOR_CLOSE") . '</button>'
						. '<button class="btn btn-success" data-dismiss="modal" aria-hidden="true" onclick="jQuery(\'#plugin' . $this->redirectPluginId . 'Modal iframe\').contents().find(\'#saveBtn\').click();">'
						. JText::_("JSAVE") . '</button>'
				)
			); ?>
		<?php endif; ?>
```
after line 53

But it does not look very good :)

![screen shot 2017-07-03 at 12 07 04](https://user-images.githubusercontent.com/869724/27788248-2bb94838-5fe8-11e7-9e0c-c0ca85a399f1.png)

roland-d committed:
Fixed the URLs for Hathor

roland-d commented:
@infograf768 I agree, we shouldn't destroy what is working ;) Instead of keeping the link you gave I did update the link to go directly to the edit screen as this also works in the isis template. Both links are the same now and not using a modal.

infograf768 commented:
You forgot
> The link in the table footer is wrong. It should be a sprintf
>line 150 to 152 should be

this is used when the plugin is disabled. We need the 
```

		<?php else : ?>
			<span class="disabled"><?php echo JText::sprintf('COM_REDIRECT_PLUGIN_DISABLED', 'index.php?option=com_plugins&filter_search=redirect'); ?></span>
		<?php endif; ?>
```
Or use as link
`index.php?option=com_plugins&task=plugin.edit&extension_id=' . RedirectHelper::getRedirectPluginId()`

as other wise we get this type of url:
`http://mysite.com/administrator/%s`

roland-d committed:
More Hathor massaging

roland-d commented:
@infograf768 Thanks, updated things again and checked and checked but who knows what else I forgot :)

Quy commented:
The sentence is repeated.

> The Redirect Plugin is enabled. The Redirect Plugin is enabled. The option 'Collect URLs' is enabled. `

infograf768 commented:
@roland-d 
we also have to take care of RTL.
I suggest this

```
	<p class="footer-tip">
		<?php if ($this->enabled && $this->collect_urls_enabled) : ?>
			<span class="enabled"><?php echo JText::sprintf('COM_REDIRECT_COLLECT_URLS_ENABLED', JText::_('COM_REDIRECT_PLUGIN_ENABLED')); ?></span>
		<?php elseif ($this->enabled && !$this->collect_urls_enabled) : ?>
			<span class="enabled"><?php echo JText::sprintf('COM_REDIRECT_COLLECT_MODAL_URLS_DISABLED', JText::_('COM_REDIRECT_PLUGIN_ENABLED'), 'index.php?option=com_plugins&task=plugin.edit&extension_id=' . RedirectHelper::getRedirectPluginId()); ?></span>
		<?php elseif (!$this->enabled) : ?>
			<span class="disabled"><?php echo JText::sprintf('COM_REDIRECT_PLUGIN_DISABLED', 'index.php?option=com_plugins&task=plugin.edit&extension_id=' . RedirectHelper::getRedirectPluginId()); ?></span>
		<?php endif; ?>
	</p>
```

infograf768 commented:
hmm, still a small issue. 
we get 
The Redirect Plugin is enabled. The 'Collect URLs' option in the index.php?option=com_plugins&task=plugin.edit&extension_id=427 is disabled. Error page URLs will not be collected by this component.
coming back with corrected proposal

infograf768 commented:
Here it is:
```
	<p class="footer-tip">
		<?php if ($this->enabled && $this->collect_urls_enabled) : ?>
			<span class="enabled"><?php echo JText::sprintf('COM_REDIRECT_COLLECT_URLS_ENABLED', JText::_('COM_REDIRECT_PLUGIN_ENABLED')); ?></span>
		<?php elseif ($this->enabled && !$this->collect_urls_enabled) : ?>
			<?php $link = JHtml::_(
					'link',
					JRoute::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . RedirectHelper::getRedirectPluginId()),
					JText::_('COM_REDIRECT_SYSTEM_PLUGIN')
				);
			?>
			<span class="enabled"><?php echo JText::sprintf('COM_REDIRECT_COLLECT_MODAL_URLS_DISABLED', JText::_('COM_REDIRECT_PLUGIN_ENABLED'), $link); ?></span>
		<?php elseif (!$this->enabled) : ?>
			<span class="disabled"><?php echo JText::sprintf('COM_REDIRECT_PLUGIN_DISABLED', 'index.php?option=com_plugins&task=plugin.edit&extension_id=' . RedirectHelper::getRedirectPluginId()); ?></span>
		<?php endif; ?>
	</p>
```
this should work fine now.

roland-d commented:
Thank you @infograf768 I have added the code verbatim as I don't have time to test it further at this moment but at least if it is all good others can test as well.

franz-wohlkoenig commented:
I have tested this item :white_check_mark: successfully on fc25aa7d1395abab1b0d7191f786cca8ff883126<br /><br />Tested Hathor, Isis; English, Persian.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

infograf768 commented:
I have tested this item :white_check_mark: successfully on fc25aa7d1395abab1b0d7191f786cca8ff883126<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

infograf768 commented:
RTC. Thanks!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16844">issues.joomla.org/tracker/joomla-cms/16844</a>.</sub>

infograf768 commented:
@rdeutz
did not set milestone. I guess it can go into 3.7.4. as we can in further releases use the new modal for other places in admin.

brianteeman commented:
As its a new feature it should be 3.8 surely

infograf768 commented:
@roland-d 
Just remarked that you are deleting in this PR
`administrator/manifests/packages/pkg_weblinks.xml`

I guess it is just an error. Please correct.

roland-d commented:
@infograf768 That is a mistake, for some reason this thing keeps popping up. I don't think this file is even part of the core. I will fix this.

roland-d committed:
RTL support in Hathor #16844 Redo

infograf768 commented:
it IS part of core ;)

infograf768 commented:
Looks like it solved itself, weird.

roland-d commented:
@infograf768 Still part huh? :/ No, it didn't solve itself, I removed the previous commit and made a new commit.

infograf768 commented:
Then git is weird as, normally, a new commit would take off the RTC label...
Anyhow, all is fine.

mbabker commented:
Needs to be synced with staging since the plugin helper class has now moved to its namespaced location (should just be able to `git merge staging` and be good, I had no issues testing another more complex PR).

roland-d committed:
Merge branch 'staging' into redo-7259-modal-for-quicker-edit

* staging: (274 commits)
  Add JCryptCipherSodium to support libsodium (#16754)
  Performance 2 (libraries/legacy) (#12220)
  Performance 6 (templates) (#12233)
  Fixed typehint (#16425)
  Fix for: Repeatable field is no longer rendered with Chosen layout (#16471)
  Fix the path for the ajax-loader.gif (#16701)
  Menu items list parent filter (#17060)
  Text Filters layout (#17113)
  mod_login showon option (#17153)
  com_banners incorret tooltip (#17157)
  fix joomla.content.options_default (#17123)
  remove the never working limitstart call (#17184)
  Update phpDocumentor build
  set 3.8.0 Dev State
  Prepare 3.7.4 Stable Release
  fixed a logic change in #12294, thanks @Hoffi1
  Update sv-SE.ini
  Update pt-BR.ini
  Update lv-LV.ini
  Update fa-IR.ini
  ...

roland-d commented:
@mbabker Done as requested

brianteeman commented:
thank you
